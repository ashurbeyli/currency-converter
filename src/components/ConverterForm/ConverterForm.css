form {
  padding: 20px 20px;
}
.exchangeSubmitArea {
  padding: 40px 0;
}