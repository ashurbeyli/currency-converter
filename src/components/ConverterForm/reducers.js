// eslint-disable-next-line import/prefer-default-export
export const balanceReducer = (state, action) => {
    switch (action.type) {
        case 'DECREASE':
            return {...state, [action.currency]: state[action.currency] - action.by};
        case 'INCREASE':
            return {...state, [action.currency]: state[action.currency] + action.by};
        default:
            throw new Error();
    }
};