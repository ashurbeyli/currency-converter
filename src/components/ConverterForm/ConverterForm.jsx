import React, { useEffect, useState, useReducer } from 'react';
import { Form, Button, Col } from 'react-bootstrap';
import _ from 'lodash';
import './ConverterForm.css';
import CurrencyControls from './CurrencyControls/CurrencyControls';
import PocketInfo from './PocketInfo/PocketInfo';
import ExchangeRate from './ExchangeRate/ExchangeRate';
import {
  convertAmount,
  getFirstFromOmittedPockets,
  omitSelectedFromPockets
} from '../../utils/pocketsUtils';

import pockets from '../../mock/data/pockets';
import mockBalance from '../../mock/data/balance';
import twoDecimalsFormatter from '../../utils/formatters/twoDecimalsFormatter';
import { balanceReducer } from './reducers';

const ConverterForm = ({ rates }) => {
  const [balance, balanceDispatcher] = useReducer(balanceReducer, mockBalance);
  const [currencyFrom, setCurrencyFrom] = useState(pockets[0]);
  const [currencyTo, setCurrencyTo] = useState(pockets[1]);
  const [fromNumber, setFromNumber] = useState('');
  const [toNumber, setToNumber] = useState('');
  const [lastUpdatedInput, setLastUpdatedInput] = useState();
  const balanceFrom = Number(_.get(balance, currencyFrom) || 0);
  const balanceTo = Number(_.get(balance, currencyTo) || 0);

  useEffect(() => {
    if (lastUpdatedInput === 'to') {
      setFromNumber(convertAmount(toNumber, currencyTo, currencyFrom, rates));
    } else {
      setToNumber(convertAmount(fromNumber, currencyFrom, currencyTo, rates));
    }
  }, [
    lastUpdatedInput,
    toNumber,
    setToNumber,
    fromNumber,
    currencyFrom,
    currencyTo,
    rates
  ]);

  const onChangeCurrencyFromSelect = evt => {
    const { value } = evt.target;
    setCurrencyFrom(value);
    setCurrencyTo(getFirstFromOmittedPockets(pockets, value));
  };

  const onCurrencyFromInputChange = value => {
    const converted = convertAmount(value, currencyFrom, currencyTo, rates);
    setFromNumber(value);
    setToNumber(converted || '');
    setLastUpdatedInput('from');
  };

  const onCurrencyToInputChange = value => {
    const converted = convertAmount(value, currencyTo, currencyFrom, rates);
    setToNumber(value);
    setFromNumber(converted || '');
    setLastUpdatedInput('to');
  };

  const checkBalanceIsValid = () =>
    Number(balanceFrom || 0) - Number(fromNumber || 0) >= 0;

  const handleSubmit = event => {
    event.preventDefault();
    const form = event.currentTarget;
    if (form.checkValidity() === false) {
      event.stopPropagation();
    }

    balanceDispatcher({
      type: 'DECREASE',
      by: Number(fromNumber || 0),
      currency: currencyFrom
    });
    balanceDispatcher({
      type: 'INCREASE',
      by: Number(toNumber || 0),
      currency: currencyTo
    });
  };

  return (
    <Form onSubmit={handleSubmit}>
      <Form.Row>
        <Col xs={5}>
          <CurrencyControls
            label="from"
            options={pockets}
            onChangeSelect={onChangeCurrencyFromSelect}
            onChangeInput={onCurrencyFromInputChange}
            inputFormatter={twoDecimalsFormatter}
            inputValue={fromNumber || ''}
          />
        </Col>
        <Col xs={2} className="text-center">
          {'=>'}
        </Col>
        <Col xs={5}>
          <CurrencyControls
            label="to"
            selected={currencyTo}
            options={omitSelectedFromPockets(pockets, currencyFrom)}
            onChangeSelect={evt => setCurrencyTo(evt.target.value)}
            onChangeInput={onCurrencyToInputChange}
            inputFormatter={twoDecimalsFormatter}
            inputValue={toNumber || ''}
          />
        </Col>
      </Form.Row>
      <Form.Row>
        <Col xs={3}>
          <PocketInfo
            currency={currencyFrom}
            balance={balanceFrom}
            value={fromNumber}
          />
        </Col>
        <Col xs={6} className="text-center">
          <ExchangeRate from={currencyFrom} to={currencyTo} rates={rates} />
        </Col>
        <Col xs={3} className="text-right">
          <PocketInfo currency={currencyTo} balance={balanceTo} />
        </Col>
      </Form.Row>
      <Form.Row>
        <Col className="text-right exchangeSubmitArea">
          <Button
            type="submit"
            variant="primary"
            size="lg"
            block
            disabled={!checkBalanceIsValid()}
          >
            Exchange
          </Button>
        </Col>
      </Form.Row>
    </Form>
  );
};

export default ConverterForm;
