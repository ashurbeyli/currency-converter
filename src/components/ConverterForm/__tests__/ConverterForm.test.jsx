import React from 'react';
import { shallow } from 'enzyme';
import ConverterForm from '../ConverterForm';

describe('<ConverterForm/>', () => {
  test('snapshot should match', () => {
    const wrapper = shallow(<ConverterForm rates={['USD', 'AZN', 'EUR']} />);
    expect(wrapper).toMatchSnapshot();
  });
});
