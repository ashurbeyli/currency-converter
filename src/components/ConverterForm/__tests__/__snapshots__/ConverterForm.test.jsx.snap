// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConverterForm/> snapshot should match 1`] = `
<Form
  inline={false}
  onSubmit={[Function]}
>
  <FormRow>
    <Col
      xs={5}
    >
      <CurrencyControls
        inputFormatter={[Function]}
        inputValue=""
        label="from"
        onChangeInput={[Function]}
        onChangeSelect={[Function]}
        options={
          Array [
            "USD",
            "EUR",
            "GBP",
          ]
        }
      />
    </Col>
    <Col
      xs={
        Object {
          "offset": 2,
          "span": 5,
        }
      }
    >
      <CurrencyControls
        inputFormatter={[Function]}
        inputValue=""
        label="to"
        onChangeInput={[Function]}
        onChangeSelect={[Function]}
        options={
          Array [
            "EUR",
            "GBP",
          ]
        }
        selected="EUR"
      />
    </Col>
  </FormRow>
  <FormRow>
    <Col
      xs={3}
    >
      <PocketInfo
        balance={100}
        currency="USD"
        value=""
      />
    </Col>
    <Col
      className="text-center"
      xs={6}
    >
      <ExchangeRate
        from="USD"
        rates={
          Array [
            "USD",
            "AZN",
            "EUR",
          ]
        }
        to="EUR"
      />
    </Col>
    <Col
      className="text-right"
      xs={3}
    >
      <PocketInfo
        balance={231}
        currency="EUR"
      />
    </Col>
  </FormRow>
  <FormRow>
    <Col
      className="text-right exchangeSubmitArea"
    >
      <Button
        active={false}
        block={true}
        disabled={false}
        size="lg"
        type="submit"
        variant="primary"
      >
        Exchange
      </Button>
    </Col>
  </FormRow>
</Form>
`;
